libass/ass_outline.c

```diff
@@ -24,45 +24,75 @@
 
 
 
-bool outline_alloc(ASS_Outline *outline, size_t n_points, size_t n_segments)
+/*
+ * \brief Initialize ASS_Outline to an empty state
+ * Equivalent to zeroing of outline object and doesn't free any memory.
+ */
+void outline_clear(ASS_Outline *outline)
 {
-    outline->points = malloc(sizeof(ASS_Vector) * n_points);
-    outline->segments = malloc(n_segments);
+    outline->points = NULL;
+    outline->segments = NULL;
+
+    outline->n_points = outline->max_points = 0;
+    outline->n_segments = outline->max_segments = 0;
+}
+
+/*
+ * \brief Initialize ASS_Outline and allocate memory
+ */
+bool outline_alloc(ASS_Outline *outline, size_t max_points, size_t max_segments)
+{
+    assert(max_points && max_segments);
+    outline->points = malloc(sizeof(ASS_Vector) * max_points);
```

Comment:
Pre-existing issue, but this will allocated to few memory if `max_points > SIZE_MAX / sizeof(ASS_Vector)`, causing subsequentmemory- overreads.
Using one of `ASS_REALLOC_ARRAY` or `ass_(try_)realloc_array` would avoid that by returning `NULL` or not modifying the (`NULL`-initialised) pointer instead.

Reply:
In all use cases `max_points` is up to factor of 2 larger than previously existing point array size. So it's physically impossible to hit that limit. I can add `assert(max_points <= SIZE_MAX / sizeof(ASS_Vector));` to catch incorrect input due to overflows, but real handling of that case looks like overkill for me.

Reply:
Is this previously existing point array size guaranteed to always be `< SIZE_MAX / sizeof(ASS_Vector)`? Regardless, it might still be good to have such an overflow check, assert or a comment addressing why it can't happen to aid readability.

Assuming this is never supposed to happen, I don't see much of a difference between an manual `assert` crashing and using `ass_realloc_array` failing the allocation. I'd be fine with either one and would prefer it over a silent assumption of overflow being impossible.

Reply:
Added a check for reliability.